### PR TITLE
Scripts for downloading models in DNN samples

### PR DESCRIPTION
**Pull request by sl-sergei**
Relates: https://github.com/opencv/opencv/pull/12186

<cut/>

```
force_builders=linux,docs,Custom,Custom Mac
build_image:Custom=ubuntu-openvino-2020.4.0:16.04
build_image:Custom Win=openvino-2020.4.0
build_image:Custom Mac=openvino-2020.4.0

test_modules:Custom=dnn,python2,python3,java,samples
test_modules:Custom Win=dnn,python2,python3,java,samples
test_modules:Custom Mac=dnn,python2,python3,java,samples

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by sl-sergei** [comment]
@alalek Can you please review these changes? Please, let me know, if there is something else needed for this task.

**Comment by asmorkalov** [comment]
@vpisarev please take a look.

**Comment by vpisarev** [comment]
@sl-sergei, thank you!

1.  can you please put force_builders=... comment into the pull request to run it one some build slaves?
2. I would like to have some decentralized solution (when all the info about downloadable data is put into one big file), and not necessarily limited to deep learning models. It can also be images, video clips and generally arbitrary data that is needed for a sample or an app. There should be documented program API as well, not just command-line API.

Approximate usage:

```
import cv2 as cv, sys

try:
   imgpath=cv.download(
            url="https://github.com/opencv/opencv/raw/master/samples/data/lena.jpg",
            sha="12345",
            save_as="lena.jpg"
            )
except IOError:
   sys.exit("could not load lena image")
# url and sha are required parameters
# save_as is optional
# dir/save_dir is another optional parameter; if it's not set,
#     some environment variable should be used (like OPENCV_SAVE_DIR),
#     and if it's empty, it could be opencv_data subdirectory inside $TEMP or something like that.
#     saving to current directory may not be very good idea, because sample can be executed from different directories
#     saving to the same directory as sample is not a good idea either,
#     because it can be in read-only directory (like /usr/share) or on a slow flash drive etc.
# maybe there should be "timeout" optional parameter.

img = cv.imread(imgpath, -1)
... 
```

**Comment by sl-sergei** [comment]
@vpisarev @alalek Can you please review the changes?

**Comment by alalek** [comment]
> cv.download

This means that this functionality should be contributed into Python package with OpenCV itself (somewhere [here](https://github.com/opencv/opencv/tree/master/modules/python/package/cv2))

---

> subdirectory inside $TEMP

There are several variants which are not secure (due to be globally writable locations, like /tmp).
Consider checking (reusing / adopting) code for OpenCL binaries cache (review this [code](https://github.com/opencv/opencv/blame/4.5.0/modules/core/src/utils/filesystem.cpp#L430)).

**Comment by sl-sergei** [comment]
> > cv.download
> 
> This means that this functionality should be contributed into Python package with OpenCV itself (somewhere [here](https://github.com/opencv/opencv/tree/master/modules/python/package/cv2))
> 
> > subdirectory inside $TEMP
> 
> There are several variants which are not secure (due to be globally writable locations, like /tmp).
> Consider checking (reusing / adopting) code for OpenCL binaries cache (review this [code](https://github.com/opencv/opencv/blame/4.5.0/modules/core/src/utils/filesystem.cpp#L430)).

Okay, I will do it

**Comment by vpisarev** [comment]
@sl-sergei, thank you! I still do not see how this functionality can be used from another Python script, as in the sample that I put above. Can you, please, modify at least one Python sample to download some data automatically?

**Comment by sl-sergei** [comment]
@alalek I updated script logic to make cache folder selection more secure. @vpisarev I added an example to README

**Comment by vpisarev** [comment]
@sl-sergei, looks good to me! :+1:

**Review by vpisarev — APPROVED** [review]

**Comment by asmorkalov** [comment]
@alalek Could we merge it?

**Review by alalek — COMMENTED** [review]
I don't see changes in current samples.
Is there some "single-click" sample? Which can download all necessary data without passing of extra parameters?

**Comment by sl-sergei** [comment]
> I don't see changes in current samples.
> Is there some "single-click" sample? Which can download all necessary data without passing of extra parameters?

```python download_models.py``` command will parse models.yml and face_detector/weights.meta4 and download all models mentioned there

**Comment by alalek** [comment]
@vpisarev @asmorkalov @sl-sergei Please provide feedback on comments above.

**Comment by sl-sergei** [comment]
@alalek I have already fixed naming for SHA1 key locally and plan to reuse getCacheDirectory, just in the middle of process

**Comment by vpisarev** [comment]
:+1:

File: samples/dnn/README.md [hunk 1]
```diff
@@ -19,6 +19,36 @@ Check `-h` option to know which values are used by default:
...
+filepath1 = downloadFile("https://drive.google.com/uc?export=download&id=0B3gersZ2cHIxRm5PMWRoTkdHdHc", None, filename="MobileNetSSD_deploy.caffemodel", save_dir="save_dir_1")
+filepath2 = downloadFile("https://drive.google.com/uc?export=download&id=0B3gersZ2cHIxRm5PMWRoTkdHdHc", "994d30a8afaa9e754d17d2373b2d62a7dfbaaf7a", filename="MobileNetSSD_deploy.caffemodel")
+print(filepath1)
+print(filepath2)
+# Your code
+```
+
+By running the following commands, you will get **MobileNetSSD_deploy.caffemodel** file:
+```bash
+export OPENCV_SAVE_DIR=save_dir_2
```
Comment (alalek) [inline samples/dnn/README.md:46]: Lets introduce more specific name here.

It is unclear what is going to be saved here. May be test results?

Something like:
- `OPENCV_DOWNLOAD_DATA_PATH`
- `OPENCV_DOWNLOAD_CACHE_DATA_PATH`


P.S. `_DATA_PATH` is aligned with `OPENCV_TEST_DATA_PATH`

File: samples/dnn/download_models.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    env_path = os.environ.get("OPENCV_SAVE_DIR", None)
+    if env_path:
+        save_dir = env_path
+    else:
+        if platform.system() == "Darwin":
+            #On Apple devices
+            temp_env = os.environ.get("TMPDIR", None)
+            if temp_env is None or not os.path.isdir(temp_env):
+                temp_dir = Path("/tmp")
+                print("Using world accessible cache directory. This may be not secure: ", temp_dir)
```
[the remark this answers is not in the document]
Comment (alalek) [reply samples/dnn/download_models.py]: Lets rollback changes and add "TODO" here until we have resolved issues in #18931

**Review by alalek — COMMENTED** [review]
> samples/dnn/download_models.py

Probably its name is very specific. We may want to download other files (like annotations, datasets) too.

Also we can keep this "as is" and then proxy into `cv.downloader` from package when it will be available.

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**